core/scmi-server/sub.mk

```diff
@@ -14,6 +14,7 @@ incdirs_ext-y += $(scmi-server-scp-path)/arch/none/optee/include
 scp-firmware-output = $(scmi-server-out-path)/build/product/optee-$(scmi-server-product)/fw/libscmi-fw-all.a
 
 libdeps += $(scp-firmware-output)
+libdeps += out/arm/core-lib/libutils/libutils.a
```

Comment:
You don't want to hard-code the output directory here! It is the reason of the CI failure, which uses a custom output directory ([link](https://github.com/OP-TEE/optee_os/actions/runs/3542522124/jobs/5948110592#step:5:183)).

The fix should be simple enough though. I suggest to define `lib-libfile-$(libname)-$(sm)` similarly to the `-shlibfile-` version which already exists:
```diff
diff --git a/mk/lib.mk b/mk/lib.mk
index 30f2040a3..65b575ec5 100644
--- a/mk/lib.mk
+++ b/mk/lib.mk
@@ -27,6 +27,7 @@ endif
 include mk/compile.mk

 lib-libfile    := $(out-dir)/$(base-prefix)$(libdir)/lib$(libname).a
+lib-libfile-$(libname)-$(sm) := $(lib-libfile)
 ifeq ($(CFG_ULIBS_SHARED),y)
 lib-shlibfile  := $(out-dir)/$(base-prefix)$(libdir)/lib$(libname).so
 lib-shlibstrippedfile := $(out-dir)/$(base-prefix)$(libdir)/lib$(libname).stripped.so
```

Then you should be able to use: `libdeps += $(lib-libfile-libutils-core)` here.

Reply:
> EDIT: But why is this dependency even needed? libscmi-fw-all.a is an archive, it should not need to be linked against libutils.

Actually... you're right, we don't need it. Libutils is already in the system deps, as the scim-server lib we're building. I'll crosscheck with @vingu-linaro, maybe related to a specific build env.

Reply:
It's not about linking with archive but the order of the lib when optee_os is linked into the final binary.
Without this dependency, libutils.a is before libscmi-fw-all.a some functions of libutils are removed because not used at this point. With this, we ensure that libscmi-fw-all.a is before and add a ref to libutils.a function

Reply:
> It's not about linking with archive but the order of the lib when optee_os is linked into the final binary. Without this dependency, libutils.a is before libscmi-fw-all.a some functions of libutils are removed because not used at this point. With this, we ensure that libscmi-fw-all.a is before and add a ref to libutils.a function

That is surprising, `libutils.a` is usually the last lib (expect `libgcc.a`) on the command line.

Reply:
> It's not about linking with archive but the order of the lib when optee_os is linked into the final binary.

Right, I was a bit confused by the name (`libdeps`), as if the library was needed to produce the SCP .a. It's not -- very well.

> That is surprising, `libutils.a` is usually the last lib (except `libgcc.a`) on the command line.

Correct, this is supposed to be achieved by:
```
mk/lib.mk:libdeps               := $(lib-libfile) $(libdeps)
```
...and the fact that `libutils` is listed first in `core.mk`:

```
# Do libraries
...
libname = utils
libdir = lib/libutils
include mk/lib.mk 
...
ifeq ($(CFG_SCMI_SERVER),y)
libname = scmi-server
libdir = core/scmi-server
include mk/lib.mk
endif
```

Reply:
without this line i have : 
/usr/bin/ccache /home/vingu/Linaro/SCMI/optee-project-qemu-latest/build/../toolchains/aarch64/bin/aarch64-linux-gnu-ld.bfd -T core/arch/arm/kernel/link_dummy.ld --no-check-sections  out/arm/core/arch/arm/kernel/ldelf_loader.o out/arm/core/arch/arm/kernel/tee_time.o out/arm/core/arch/arm/kernel/rpc_io_i2c.o out/arm/core/arch/arm/kernel/otp_stubs.o out/arm/core/arch/arm/kernel/delay.o out/arm/core/arch/arm/kernel/idle.o out/arm/core/arch/arm/kernel/tee_time_arm_cntpct.o out/arm/core/arch/arm/kernel/timer_a64.o out/arm/core/arch/arm/kernel/spin_lock_a64.o out/arm/core/arch/arm/kernel/tlb_helpers_a64.o out/arm/core/arch/arm/kernel/cache_helpers_a64.o out/arm/core/arch/arm/kernel/thread_a64.o out/arm/core/arch/arm/kernel/thread.o out/arm/core/arch/arm/kernel/thread_optee_smc.o out/arm/core/arch/arm/kernel/thread_optee_smc_a64.o out/arm/core/arch/arm/kernel/abort.o out/arm/core/arch/arm/kernel/vfp.o out/arm/core/arch/arm/kernel/vfp_a64.o out/arm/core/arch/arm/kernel/trace_ext.o out/arm/core/arch/arm/kernel/misc_a64.o out/arm/core/arch/arm/kernel/boot.o out/arm/core/arch/arm/kernel/entry_a64.o out/arm/core/arch/arm/kernel/unwind_arm64.o out/arm/core/arch/arm/mm/core_mmu.o out/arm/core/arch/arm/mm/core_mmu_lpae.o out/arm/core/arch/arm/mm/pgt_cache.o out/arm/core/arch/arm/mm/mobj_dyn_shm.o out/arm/core/arch/arm/tee/arch_svc_a64.o out/arm/core/arch/arm/tee/arch_svc.o out/arm/core/arch/arm/tee/entry_fast.o out/arm/core/arch/arm/tee/cache.o out/arm/core/arch/arm/plat-vexpress/main.o out/arm/core/arch/arm/plat-vexpress/vendor_props.o out/arm/core/crypto/crypto.o out/arm/core/crypto/aes-gcm.o out/arm/core/crypto/aes-gcm-sw.o out/arm/core/crypto/aes-gcm-ghash-tbl.o out/arm/core/crypto/signed_hdr.o out/arm/core/crypto/rng_fortuna.o out/arm/core/crypto/cbc-mac.o out/arm/core/crypto/aes-cts.o out/arm/core/crypto/sm2-kdf.o out/arm/core/crypto/sm3.o out/arm/core/crypto/sm3-hash.o out/arm/core/crypto/sm3-hmac.o out/arm/core/crypto/sm4.o out/arm/core/crypto/sm4-ecb.o out/arm/core/crypto/sm4-cbc.o out/arm/core/crypto/sm4-ctr.o out/arm/core/drivers/pl011.o out/arm/core/drivers/gic.o out/arm/core/drivers/clk/clk.o out/arm/core/drivers/clk/clk_dt.o out/arm/core/drivers/clk/fixed_clk.o out/arm/core/drivers/rstctrl/rstctrl.o out/arm/core/kernel/spin_lock_debug.o out/arm/core/kernel/assert.o out/arm/core/kernel/console.o out/arm/core/kernel/dt.o out/arm/core/kernel/dt_driver.o out/arm/core/kernel/dt_driver_test.o out/arm/core/kernel/pm.o out/arm/core/kernel/handle.o out/arm/core/kernel/interrupt.o out/arm/core/kernel/ldelf_syscalls.o out/arm/core/kernel/msg_param.o out/arm/core/kernel/panic.o out/arm/core/kernel/refcount.o out/arm/core/kernel/tee_misc.o out/arm/core/kernel/tee_ta_manager.o out/arm/core/kernel/ts_manager.o out/arm/core/kernel/scattered_array.o out/arm/core/kernel/huk_subkey.o out/arm/core/kernel/user_mode_ctx.o out/arm/core/kernel/initcall.o out/arm/core/kernel/user_access.o out/arm/core/kernel/mutex.o out/arm/core/kernel/wait_queue.o out/arm/core/kernel/notif.o out/arm/core/kernel/thread.o out/arm/core/kernel/user_ta.o out/arm/core/kernel/ree_fs_ta.o out/arm/core/kernel/early_ta.o out/arm/core/kernel/secstor_ta.o out/arm/core/kernel/embedded_ts.o out/arm/core/kernel/pseudo_ta.o out/arm/core/mm/mobj.o out/arm/core/mm/fobj.o out/arm/core/mm/file.o out/arm/core/mm/vm.o out/arm/core/mm/core_mmu.o out/arm/core/mm/tee_mm.o out/arm/core/pta/device.o out/arm/core/pta/secstor_ta_mgmt.o out/arm/core/pta/stats.o out/arm/core/pta/system.o out/arm/core/pta/scmi.o out/arm/core/pta/tests/fs_htree.o out/arm/core/pta/tests/invoke.o out/arm/core/pta/tests/misc.o out/arm/core/pta/tests/mutex.o out/arm/core/pta/tests/aes_perf.o out/arm/core/tee/entry_std.o out/arm/core/tee/tee_cryp_utl.o out/arm/core/tee/tee_cryp_hkdf.o out/arm/core/tee/tee_cryp_concat_kdf.o out/arm/core/tee/tee_cryp_pbkdf2.o out/arm/core/tee/tee_obj.o out/arm/core/tee/tee_svc.o out/arm/core/tee/tee_svc_cryp.o out/arm/core/tee/tee_svc_storage.o out/arm/core/tee/tee_time_generic.o out/arm/core/tee/tadb.o out/arm/core/tee/socket.o out/arm/core/tee/tee_ta_enc_manager.o out/arm/core/tee/tee_fs_key_manager.o out/arm/core/tee/tee_ree_fs.o out/arm/core/tee/fs_dirfile.o out/arm/core/tee/fs_htree.o out/arm/core/tee/tee_fs_rpc.o out/arm/core/tee/tee_pobj.o out/arm/core/tee/uuid.o out/arm/core/tee/tee_supp_plugin_rpc.o out/arm/core/ta_pub_key.o out/arm/core/ldelf_hex.o out/arm/core/early_ta_f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c.o out/arm/core/arch/arm/dts/embedded_dtb_test.o   out/arm/core/scmi-server/libscmi-server.a out/arm/lib/libunw/libunw.a out/arm/core/lib/zlib/libzlib.a out/arm/core/lib/libfdt/libfdt.a out/arm/core/lib/libtomcrypt/libtomcrypt.a out/arm/core-lib/libmbedtls/libmbedtls.a out/arm/core-lib/libutils/libutils.a  out/arm/core/scmi-server/build/product/optee-fvp/fw/libscmi-fw-all.a /home/vingu/Linaro/SCMI/optee-project-qemu-latest/toolchains/aarch64/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/libgcc.a -o out/arm/core/all_objs.o
/home/vingu/Linaro/SCMI/optee-project-qemu-latest/build/../toolchains/aarch64/bin/aarch64-linux-gnu-ld.bfd: /home/vingu/Linaro/SCMI/optee-project-qemu-latest/optee_os/out/arm/core/scmi-server/build/framework/libframework.a(fwk_string.c.obj): in function `fwk_str_strncpy':
/home/vingu/Linaro/SCMI/optee-project-qemu-latest/SCP-firmware/framework/src/fwk_string.c:52: undefined reference to `strncpy'
core/arch/arm/kernel/link.mk:68: recipe for target 'out/arm/core/all_objs.o' failed

With the line it becomes: 

/usr/bin/ccache /home/vingu/Linaro/SCMI/optee-project-qemu-latest/build/../toolchains/aarch64/bin/aarch64-linux-gnu-ld.bfd -T core/arch/arm/kernel/link_dummy.ld --no-check-sections  out/arm/core/arch/arm/kernel/ldelf_loader.o out/arm/core/arch/arm/kernel/tee_time.o out/arm/core/arch/arm/kernel/rpc_io_i2c.o out/arm/core/arch/arm/kernel/otp_stubs.o out/arm/core/arch/arm/kernel/delay.o out/arm/core/arch/arm/kernel/idle.o out/arm/core/arch/arm/kernel/tee_time_arm_cntpct.o out/arm/core/arch/arm/kernel/timer_a64.o out/arm/core/arch/arm/kernel/spin_lock_a64.o out/arm/core/arch/arm/kernel/tlb_helpers_a64.o out/arm/core/arch/arm/kernel/cache_helpers_a64.o out/arm/core/arch/arm/kernel/thread_a64.o out/arm/core/arch/arm/kernel/thread.o out/arm/core/arch/arm/kernel/thread_optee_smc.o out/arm/core/arch/arm/kernel/thread_optee_smc_a64.o out/arm/core/arch/arm/kernel/abort.o out/arm/core/arch/arm/kernel/vfp.o out/arm/core/arch/arm/kernel/vfp_a64.o out/arm/core/arch/arm/kernel/trace_ext.o out/arm/core/arch/arm/kernel/misc_a64.o out/arm/core/arch/arm/kernel/boot.o out/arm/core/arch/arm/kernel/entry_a64.o out/arm/core/arch/arm/kernel/unwind_arm64.o out/arm/core/arch/arm/mm/core_mmu.o out/arm/core/arch/arm/mm/core_mmu_lpae.o out/arm/core/arch/arm/mm/pgt_cache.o out/arm/core/arch/arm/mm/mobj_dyn_shm.o out/arm/core/arch/arm/tee/arch_svc_a64.o out/arm/core/arch/arm/tee/arch_svc.o out/arm/core/arch/arm/tee/entry_fast.o out/arm/core/arch/arm/tee/cache.o out/arm/core/arch/arm/plat-vexpress/main.o out/arm/core/arch/arm/plat-vexpress/vendor_props.o out/arm/core/crypto/crypto.o out/arm/core/crypto/aes-gcm.o out/arm/core/crypto/aes-gcm-sw.o out/arm/core/crypto/aes-gcm-ghash-tbl.o out/arm/core/crypto/signed_hdr.o out/arm/core/crypto/rng_fortuna.o out/arm/core/crypto/cbc-mac.o out/arm/core/crypto/aes-cts.o out/arm/core/crypto/sm2-kdf.o out/arm/core/crypto/sm3.o out/arm/core/crypto/sm3-hash.o out/arm/core/crypto/sm3-hmac.o out/arm/core/crypto/sm4.o out/arm/core/crypto/sm4-ecb.o out/arm/core/crypto/sm4-cbc.o out/arm/core/crypto/sm4-ctr.o out/arm/core/drivers/pl011.o out/arm/core/drivers/gic.o out/arm/core/drivers/clk/clk.o out/arm/core/drivers/clk/clk_dt.o out/arm/core/drivers/clk/fixed_clk.o out/arm/core/drivers/rstctrl/rstctrl.o out/arm/core/kernel/spin_lock_debug.o out/arm/core/kernel/assert.o out/arm/core/kernel/console.o out/arm/core/kernel/dt.o out/arm/core/kernel/dt_driver.o out/arm/core/kernel/dt_driver_test.o out/arm/core/kernel/pm.o out/arm/core/kernel/handle.o out/arm/core/kernel/interrupt.o out/arm/core/kernel/ldelf_syscalls.o out/arm/core/kernel/msg_param.o out/arm/core/kernel/panic.o out/arm/core/kernel/refcount.o out/arm/core/kernel/tee_misc.o out/arm/core/kernel/tee_ta_manager.o out/arm/core/kernel/ts_manager.o out/arm/core/kernel/scattered_array.o out/arm/core/kernel/huk_subkey.o out/arm/core/kernel/user_mode_ctx.o out/arm/core/kernel/initcall.o out/arm/core/kernel/user_access.o out/arm/core/kernel/mutex.o out/arm/core/kernel/wait_queue.o out/arm/core/kernel/notif.o out/arm/core/kernel/thread.o out/arm/core/kernel/user_ta.o out/arm/core/kernel/ree_fs_ta.o out/arm/core/kernel/early_ta.o out/arm/core/kernel/secstor_ta.o out/arm/core/kernel/embedded_ts.o out/arm/core/kernel/pseudo_ta.o out/arm/core/mm/mobj.o out/arm/core/mm/fobj.o out/arm/core/mm/file.o out/arm/core/mm/vm.o out/arm/core/mm/core_mmu.o out/arm/core/mm/tee_mm.o out/arm/core/pta/device.o out/arm/core/pta/secstor_ta_mgmt.o out/arm/core/pta/stats.o out/arm/core/pta/system.o out/arm/core/pta/scmi.o out/arm/core/pta/tests/fs_htree.o out/arm/core/pta/tests/invoke.o out/arm/core/pta/tests/misc.o out/arm/core/pta/tests/mutex.o out/arm/core/pta/tests/aes_perf.o out/arm/core/tee/entry_std.o out/arm/core/tee/tee_cryp_utl.o out/arm/core/tee/tee_cryp_hkdf.o out/arm/core/tee/tee_cryp_concat_kdf.o out/arm/core/tee/tee_cryp_pbkdf2.o out/arm/core/tee/tee_obj.o out/arm/core/tee/tee_svc.o out/arm/core/tee/tee_svc_cryp.o out/arm/core/tee/tee_svc_storage.o out/arm/core/tee/tee_time_generic.o out/arm/core/tee/tadb.o out/arm/core/tee/socket.o out/arm/core/tee/tee_ta_enc_manager.o out/arm/core/tee/tee_fs_key_manager.o out/arm/core/tee/tee_ree_fs.o out/arm/core/tee/fs_dirfile.o out/arm/core/tee/fs_htree.o out/arm/core/tee/tee_fs_rpc.o out/arm/core/tee/tee_pobj.o out/arm/core/tee/uuid.o out/arm/core/tee/tee_supp_plugin_rpc.o out/arm/core/ta_pub_key.o out/arm/core/ldelf_hex.o out/arm/core/early_ta_f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c.o out/arm/core/arch/arm/dts/embedded_dtb_test.o   out/arm/core/scmi-server/libscmi-server.a out/arm/lib/libunw/libunw.a out/arm/core/lib/zlib/libzlib.a out/arm/core/lib/libfdt/libfdt.a out/arm/core/lib/libtomcrypt/libtomcrypt.a out/arm/core-lib/libmbedtls/libmbedtls.a out/arm/core-lib/libutils/libutils.a  out/arm/core/scmi-server/build/product/optee-fvp/fw/libscmi-fw-all.a out/arm/core-lib/libutils/libutils.a /home/vingu/Linaro/SCMI/optee-project-qemu-latest/toolchains/aarch64/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/libgcc.a -o out/arm/core/all_objs.o

The only difference is the position of libscmi-fw-all.a vs libutils.a

Reply:
moving `scmi-server` before `libutils` in _core/core.mk_ would resolve the dependency. But maybe a bit fragile.

Reply:
I have also considered this change but I agree that it's a bit fragile

Reply:
> moving `scmi-server` before `libutils` in _core/core.mk_ would resolve the dependency. But maybe a bit fragile.

The libs should be in reverse order of appearance in core.mk. `mk/lib.mk` should take care of that, perhaps there is a bug in it or in the way it is used.

The naming is a bit weird too, I see no reason not stick more closely with the lib model and have all this code under `core/lib/libscmi-fwk-all`. Any deviation can only be painful when looking for stuff etc.

Reply:
The problem is one line above. `mk/lib.mk` adds to `libdeps` in front and here you're appending.

Reply:
thanks @jenswi-linaro, indeed changing line 16 should fix your issue @vingu-linaro:
```patch
-libdeps += $(scp-firmware-output)
+libdeps := $(scp-firmware-output) $(libdeps)
```

I'll push a fixup commit.